src/azure-cli/azure/cli/command_modules/role/custom.py

```diff
@@ -1745,18 +1745,28 @@ def _gen_guid():
 def _random_password(length):
     import random
     import string
-    random_source = string.ascii_letters + string.digits + string.punctuation
-    password = random.choice(string.ascii_lowercase)
-    password += random.choice(string.ascii_uppercase)
-    password += random.choice(string.digits)
-    password += random.choice(string.punctuation)
+    safe_punctuation = '-_.~'
+    random_source = string.ascii_letters + string.digits + safe_punctuation
+    alphanumeric = string.ascii_letters + string.digits
 
-    for i in range(length - 4):    # pylint: disable=unused-variable
-        password += random.choice(random_source)
+    # make sure first character is not a punctuation like '--' which will make CLI command break
+    first_character = random.SystemRandom().choice(alphanumeric)
 
+    # make sure we have special character in the password
+    password = random.SystemRandom().choice(string.ascii_lowercase)
+    password += random.SystemRandom().choice(string.ascii_uppercase)
+    password += random.SystemRandom().choice(string.digits)
+    password += random.SystemRandom().choice(safe_punctuation)
+
+    # generate a password of the given length from the options in the random_source variable
+    for _ in range(length - 5):
```

Comment:
For `_random_password` function itself, do we need to consider the case when length is less than 5 ?

Reply:
the length is 34 which is not configurable for the customer